src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace_ProjectReload.cs

```diff
@@ -0,0 +1,309 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Serialization;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MSBuild
+{
+    public sealed partial class MSBuildWorkspace
+    {
+        private async Task<ImmutableArray<Project>> UpdateProjectsAsync(ImmutableArray<ProjectInfo> newProjectInfos)
+        {
+            using (_serializationLock.DisposableWait())
+            {
+                var serialization = this.Services.GetRequiredService<ISerializerService>();
+                var updatedProjectIds = TemporaryArray<ProjectId>.Empty;
+                var oldSolution = this.CurrentSolution;
+                var solution = oldSolution;
+                foreach (var projectInfo in newProjectInfos)
+                {
+                    var projectToUpdate = solution.GetProject(projectInfo.Id);
+                    RoslynDebug.AssertNotNull(projectToUpdate);
+                    var oldCheckSum = await projectToUpdate.State.GetStateChecksumsAsync(default).ConfigureAwait(false);
+                    var newCheckSum = projectInfo.GetCheckSum(serialization);
+                    var newProject = await UpdateProjectAsync(projectToUpdate, projectInfo, oldCheckSum, newCheckSum).ConfigureAwait(false);
+                    solution = AdjustReloadedProject(projectToUpdate, newProject).Solution;
+                    updatedProjectIds.Add(newProject.Id);
+                }
+
+                this.SetCurrentSolution(solution);
+                using var _1 = ArrayBuilder<Project>.GetInstance(out var updatedProjects);
+                foreach (var projectId in updatedProjectIds)
+                {
+                    _ = this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectReloaded, oldSolution, solution, projectId);
+                    updatedProjects.Add(solution.GetProject(projectId)!);
+                }
+
+                return updatedProjects.ToImmutableAndFree();
+            }
+        }
+
+        private static async Task<Project> UpdateProjectAsync(Project project, ProjectInfo newProjectInfo, ProjectStateChecksums oldCheckSum, ProjectStateChecksums newCheckSum)
+        {
+            // changed info
+            if (oldCheckSum.Info != newCheckSum.Info)
+            {
+                project = UpdateProjectAttributes(project, newProjectInfo.Attributes);
+            }
+
+            // changed compilation options
+            if (oldCheckSum.CompilationOptions != newCheckSum.CompilationOptions)
```

Comment:
The With* methods here are already going to do checks to say if the value is unchanged to avoid creating new snapshots -- we shouldn't have to be duplicating any logic here; that should greatly simplify this code.